[체스 - 1, 2단계] 조이썬(이영수) 미션 제출합니다.

youngsu5582 opened the pull request:
제이 안녕하세요!
이번에 6기 백엔드 로 참여하게 된 조이썬 이라고 합니다!!
( 옛날에  갑자기 제이의 레벨4 [회고](https://github.com/woowacourse/woowa-writing-1/pull/187)가 갑자기 떠서 봤는데 정말 재밌어서 
리뷰어로 뜬걸 보자마자 신기했습니다 ㅋ⫬ㅋ⫬ㅋ⫬ )

이제 본격적인 코드와 리뷰에 관련된 내용 시작하겠습니다!

악명 높은 체스를 직접 만나보니 정말 어려움을 느꼈습니다...

특히 체스를 그전에 한번도 해보지 않았는데,
도메인에 대한 지식이 부족하니 
자연스레 놓치거나 생각 중에 구현의 어려움을 느껴
도메인에 대한 이해의 중요성을 깨닫게 된 거 같았습니다!

### 궁금한 점

#### 1. List<Piece> VS List<Point,Piece> 

현재 저의 도메인은
```java
public abstract class Piece implements Movable {

    protected Point point;
    protected final Color color;
```
이렇게 기물이 Point 와 Color 를 가지고 있습니다
기물이 자신의 이동 과 자신이 갈 수 있는지를 검증 해주는게
객체지향적이라고 생각해서 인데요!

이때 이 Piece 를 가지는 일급 컬렉션이 어떻게 가지는지에 대해 고민하게 되었습니다

##### Map 을 사용한다면?

-> Point 를 통해 바로 get으로 받아올 수 있다
-> Piece 의 Point 와 Map 의 Point 를 계속 갱신해야 한다

##### List 를 사용한다면?

-> Point 에 해당하는 값이 있는지 계속 순회 해야 한다
-> Key 와 Value 내 Point 가 있는게 아닌 List<Piece> 를 통해 Piece 에만 Point가 있다

이 두가지의 점을 생각했는데
성능 과 편의상 Map 이 더 좋으나
Map 의 Point 와 Piece 의 Point 일관성 관리가 어려울 거 같아서 List 를 선택했습니다!

이렇게 Key 에 있는 타입이 Value Domain 내부에 있는 경우가 타당한지에 대한 제이의 의견이 궁금합니다!

#### 2. DTO

이번 미션에서 getter , setter 를 사용하지 말라는 추가 요구 사항이 있었습니다.

이에 따라 도메인을 Controller 나 View 가 알지 못하게 하기 위해
DTO를 처음 사용해봤습니다!

이때 궁금한 점으로는 결국 콘솔 프로그램인 이상 
문자열을 통해 출력을 할 수 밖에 없습니다.

DTO 가 Enum 값을 가지는 것 역시 도메인을 안다고 생각하여
데이터를 원시값으로 가지고 있게 하는 방법을 생각했습니다.

이를 위해서 Mapper 를 만들어서
도메인을 출력 String 으로 변환을 담당했습니다

Domain 및 Entity 클래스를 가지는 것은 방지 하더라도
Enum 이나 Domain 내부 VO 에 대한 값들은 가져도 괜찮은지에 대해서 궁금합니다

( 그렇지 않으면
 너무 많은 변환이나 , 너무 많은 클래스 + Mapper 를 유발할 것 같아서 입니다! )

+ 추가적인 궁금함으로 , 
Domain 을 view 요구사항에 맞쳐 toString 을 하면 안된다고 한 내용에 대해서는 공부했습니다!
그러면 DTO 는 toString 을 재정의 하여 view 에 재정의 를 해도 괜찮을까요?? 

( DTO 는 결국 View 단으로 전달하기 위해서고
 DTO 는 Domain 및 비즈니스 로직에 영향을 안주는 점 때문에 그렇게 생각합니다! )

#### 3. Getter 를 지양함 에 따른 trade-off

이번 미션 요구사항을 따르기 위해
getter 를 통해 도메인을 가져와서 수행하는게 아닌
최대한 많은 메시지를 호출하게 만들었습니다!

```
    public boolean sameColor(final Piece piece) {...}

    public boolean isBlack() {...}

    public boolean isWhite() {...}

    public boolean isDirectionStraight(final Point point) {...}

    public boolean isDirectionDiagonal(final Point point) {...}
```
이렇게 수많은 응답을 위한 메소드들이 탄생하게 됐습니다.

이 부분에서 궁금함이 들었습니다!

Color 만 넘겨주면 isBlack , isWhite , sameColor 라는 코드는 필요없게 됩니다.
물론 , 최대한 외부를 노출하지 않고 객체 끼리 소통하는게 객체지향적 구조라고 생각하지만
Enum 의 값은 결국 정해져 있는점 + 복잡한 로직 & 변환 로직 이 없는점들을 생각해
코드 길이 및 불필요한 부분을 지울수도 있다고 생각이 들었습니다

이런 부분들을 희생해서라도 완벽한 객체지향적인 점을 가져야 할 부분이 있을까요??

---

시간이 부족하고 , 체스가 너무 어려워서
코드적 으로 놓친 부분이 되게 많은거 같습니다 😭 ( 변수명 , 함수 길이 , 테스트 코드 충분도 ...)
코드적인 부분에서 궁금한 점들은 Files Changed 남긴후 Comment 로 올리겠습니다!

youngsu5582 committed:
docs(README) : 기능 요구사항 초안 정리

youngsu5582 committed:
feat(Point) : 위치를 가지는 도메인 생성 기능 구현

- 이에 따른 테스트 코드 추가

youngsu5582 committed:
feat(Piece) : 기물 도메인 생성 기능 구현

상위 추상 클래스 Piece 아래
Bishop , King , Pawn , Knight , Queen , Rook 추가
- 이에 따른 테스트 코드 추가

youngsu5582 committed:
refactor(Point) : 위치 도메인 기물 패키지로 이동

youngsu5582 committed:
refactor(Piece) : 기물에 색깔 추가

youngsu5582 committed:
feat(Pieces) : 기물 일급 컬렉션 생성 기능 구현

- 이에 따른 테스트 코드 추가

youngsu5582 committed:
feat(ChessBoard) : 체스판 생성 기능 구현

- 이에 따른 테스트 코드 추가

youngsu5582 committed:
feat(Piece) : 비숍,킹,나이트,폰,룩,퀸 도메인 기물 상속 추가

youngsu5582 committed:
refactor(All) : 도메인들 도메인 패키지로 이동

youngsu5582 committed:
feat(ChessBoardGenerator) : 기본 체스판을 생성해주는 클래스 구현

youngsu5582 committed:
feat(ChessBoard,Pieces) : 포인트를 통해 기물을 찾는 기능 구현

Piece VO화
- 이에 따른 테스트 코드 추가

youngsu5582 committed:
feat(PieceStatus) : 자신의 상태를 알려주는 기능 추가

Queen,Rook,King,Pawn,Bishop,Knight 에도 반영

youngsu5582 committed:
test(ChesBoard) : 기본 체스판을 생성하는 기능 구현

- 이에 따른 테스트 코드 추가

youngsu5582 committed:
feat(BoardDto,BoardMapper) : DTO 변환 기능 구현

Piece 도메인에 getPoint
Point 도메인에 getFileIndex,getRankIndex
ChessBoard 도메인에 toMap 추가

youngsu5582 committed:
feat(InputView,ChessCommand) : 체스 명령어를 입력 받는 기능 구현

youngsu5582 committed:
feat(OutputView) : 체스판을 출력하는 기능 구현

- 이에 따른 테스트 코드 추가

youngsu5582 committed:
feat(All) : 1단계 기능 요구사항 완료

- 이에 따른 기능 요구사항 체크

youngsu5582 committed:
feat(Point) : 파일,랭크 를 포함한 문자열을 통해 생성 기능 구현

- 이에 따른 테스트 코드 추가

youngsu5582 committed:
feat(Point) : 파일,랭크 를 포함한 문자열을 통해 생성 기능 구현

File,Rank 에 from 정적 팩토리 메소드 추가
- 이에 따른 테스트 코드 추가

youngsu5582 committed:
feat(File,Rank) : 그전 과 다음 요소 받는 기능 구현 , 경계선 위치하는지 확인하는 기능 구현

- 이에 따른 테스트 코드 추가

youngsu5582 committed:
feat(Direction) : 방향에 따라 이동하는 기능 구현

- 이에 따른 테스트 코드 추가

youngsu5582 committed:
feat(Pawn) : 폰이 이동하는 가능한지 확인하는 기능 구현

- 이에 따른 테스트 코드 추가

youngsu5582 committed:
feat(Index) : 포인트 를 숫자로 변환하는 기능 구현

Direction 의 move 테스트도 변경
- 이에 따른 테스트 코드 추가

youngsu5582 committed:
feat(DirectionFactory) : 방향 결정해주는 기능 구현

- 이에 따른 테스트 코드 추가

youngsu5582 committed:
refactor(all) : 도메인 패키지 재 분리

youngsu5582 committed:
feat(Rook) : 룩 기물이 이동 가능한지 확인하는 기능 구현

youngsu5582 committed:
feat(Bishop) : 비숍 기물이 이동 가능한지 확인하는 기능 구현

- 이에 따른 테스트 코드 추가

youngsu5582 committed:
feat(Queen) : 퀸 기물이 이동 가능한지 확인하는 기능 구현

- 이에 따른 테스트 코드 추가

youngsu5582 committed:
feat(Index) : 인덱스가 방향에 따라 이동하는 기능 구현

- 이에 따른 테스트 코드 추가

youngsu5582 committed:
feat(Direction) : 방향이 세로(위,아래) 인지 확인하는 기능 구현

- 이에 따른 테스트 코드 추가

youngsu5582 committed:
feat(Pawn) : 폰이 이동하지 않을 시 두 칸 이동하는 기능 구현

Piece 멤버 변수 protected 로 변경
Point 가 방향 결정 담당하게 변경
- 이에 따른 테스트 코드 추가

youngsu5582 committed:
feat(King) : 킹이 이동 가능한지 확인하는 기능 구현

- 이에 따른 테스트 코드 추가

youngsu5582 committed:
feat(Knight) : 나이트가 이동 가능한지 확인 하는 기능 구현

- 이에 따른 테스트 코드 추가

youngsu5582 committed:
feat(Pieces) : 기물과 포인트에 있는 기물이 아군인지 확인하는 기능 구현

- 이에 따른 테스트 코드 추가

youngsu5582 committed:
feat(Pieces) : 포인트 와 포인트 전까지 기물이 존재하는지 확인하는 기능 구현

인덱스가 해당 인덱스까지 경로를 반환하는 기능 구현
- 이에 따른 테스트 코드 추가

youngsu5582 committed:
feat(Piece) : 기물이 흑인지 백인지 알려주는 기능 구현

youngsu5582 committed:
docs(README) : 2단계 기능 요구사항 작성

youngsu5582 committed:
feat(Movable) : 이동 가능한지 알려주는 canMove 인터페이스 추상화

기물 클래스에 반영
PieceImpl 클래스에도 반영

youngsu5582 committed:
refactor(piece) : 방향 정적 임포트로 수정

Bishop,Rook,Queen,Knight,Pawn 클래스에 반영

youngsu5582 committed:
feat(File,Rank) : 파일 , 랭크 각각 원시값을 통해 생성 구현

- 이에 따른 테스트 코드 구현

youngsu5582 committed:
feat(Pieces) : 기물이 기물들 사이 이동 가능한지 기능 구현

Direction 이 직선인지 대각선인지 확인하는 기능 추가
- 이에 따른 테스트 코드 구현

youngsu5582 committed:
feat(Pieces) : 체스기물이 이동하는 기능 구현

- 이에 따른 테스트 코드 구현

youngsu5582 committed:
feat(Pieces) : 체스기물이 이동하는 기능 구현

입출력용 RouteDTO 추가
- 이에 따른 테스트 코드 구현

youngsu5582 committed:
feat(ChessController) : 체스 게임을 진행하는 기능 구현

ChessCommand 에 MOVE 명령어 추가
- 기능 완료에 따른 요구 사항 체크

youngsu5582 committed:
refactor(Index) : list -> pathIndexes 로 변수명 변경

ChessCommand 에 MOVE 명령어 추가
- 기능 완료에 따른 요구 사항 체크

youngsu5582 committed:
refactor(All) : 도메인 패키지 재분리

youngsu5582 committed:
refactor(DirectionFactory) : 클래스 명 변경

기존 : DirectionFactory
변경 : DirectionUtil

youngsu5582 committed:
refactor(All) : 매개변수 , 지역변수 final 키워드 추가

youngsu5582 committed:
docs(README) : TODO 추가

- 도메인적 추가 구현 사항
- 도메인 외적 추가 기능 구현 사항

youngsu5582 commented on src/main/java/domain/piece/Pieces.java:
```diff
@@ -0,0 +1,76 @@
+        Map<Point, Piece> map = new HashMap<>();
+
+        for (final Piece piece : value) {
+            map.put(piece.getPoint(), piece);
```
해당 부분에서 결국 getter 를 사용했지만
DTO 를 만들기 위해서 필연적으로 발생하기에
이정도 부분에서는 사용을 해도 되는지 궁금합니다..!

JunHoPark93 replied:
넵 이정도는 괜찮습니다. 다만 dto를 만들기 위함이라고 하셨는데 Pieces 도메인이 해야하는 역할이 맞는지 고민해보시면 좋을거같아요~

youngsu5582 replied:
DTO 및 외부 전달 용 getter 는 저도 필연적이라고 생각합니다!

하지만 , 기물들 일급 컬렉션이 일급 객체 들을 굳이 Mapping 해줄 필요는 없는거 같습니다!
그대로 반환 후 , Mapper 에서 변환 로직을 담당하게 책임을 주게 변환하겠습니다. 감사합니다!!

youngsu5582 commented on src/main/java/domain/piece/attribute/point/Direction.java:
```diff
@@ -0,0 +1,33 @@
+
+import java.util.function.UnaryOperator;
+
+public enum Direction {
```
해당 Enum 에서
처음에는 숫자 두개를 한 후 
외부에서 index 를 만들까 했지만
Direction 과 Index 를 의도적으로 연결되게 만든후
로직 에서 적극적으로 활용하는 것이 
1. 연관된 로직이 Enum 에서 관리된다
2. 외부에서 불필요한 생성자 코드를 작성하지 않아도 된다

이런 의미의 접근으로 
Enum 에서 다른 외부 도메인의 생성을 담당 해도 괜찮을까요?!

JunHoPark93 replied:
enum의 역할과는 멀어지는거같아요 역할을 분리해보면 어떨까요?

youngsu5582 replied:
제이!
해당 부분에 대해서
계속 생각을 해봤습니다 ...
처음에는 , Enum 이 숫자나 연산에 필요한 값들을 가지고 있게 하는 방법을 생각했으나
이 역시도 도메인 과 객체 지향적인 코드가 아니라고 생각이 들었습니다,,

그렇다고 모든 방향이나 경우에 대해서 각 기물들 마다 별도로 로직을 넣는건 별로라고 생각해서
```java
    public boolean canMoveLeft(final int count) {
        return this.file.canMoveLeft(count);
    }
    public Point moveLeft(final int count) {
        return new Point(this.file.moveLeft(count), this.rank);
    }
```

Point 가 스스로를 검증 & 움직이게 하고
```java
LEFT(Point::moveLeft, Point::canMoveLeft),
```
Direction 에서는 Point 를 받아와서 대신 호출 하는 방식을 사용해서 코드 간결화 를 추구했습니다!
해당 코드 역시도 결국 Enum 이 외부 도메인의 로직을 담당하는 느낌일가요??

JunHoPark93 replied:
movePointFunction 까지는 Direction에서 제공해주는게 어색하진않은데 canMovePredicate까지 담당하게 되면 역할이 많아지는거같아요. 기물이 스스로 움직임을 판단하는게(실제 작성해주신 canMove메서드) 더 자연스러워보이는데 한번 고민해보시고 판단해보세요!
꼭 반영하지 않으셔도됩니다.

youngsu5582 commented on src/main/java/domain/piece/attribute/point/File.java:
```diff
@@ -0,0 +1,35 @@
+        throw new IllegalArgumentException(String.format("%c는 파일에 존재하지 않습니다.", value));
+    }
+
+    public static File findByIndex(final int ordinalIndex) {
```
해당 코드는 ordinal 에 해당하는
File 의 값을 반환해주는 코드입니다.
이 코드는 외부에서 isInBoundary 를 통해 호출후
코드를 잘못 사용할 가능성이 없다고 생각을 했습니다.

하지만 , 모르는 개발자가 사용할 경우를 대비해
if문을 호출한 후 사용하는 코드 이더라도
내부에서 
```java
if(!isInBoundary){
    throw new IllegalStateException("");
}
```
이렇게 검증을 한 후 사용을 해야할까요??

---

이런 public 로직에서 다른 public 로직을 사용해도 괜찮나요??
제가 생각하는 이유로는

외부에 노출되는 로직이
내부에서 다른 노출되는 로직의 내부에서 사용 되는 것이
이상하다고 생각해서 입니다.

그렇지만 이런 점을 방지하기 위해서
내부 함수로 래핑을 하는 것은 이상한 거 같기도 합니다..!

( 아직 기본기가 많이 부족한 거 같습니다... 🥹🥹 )

JunHoPark93 replied:
이렇게 보이지 않는 의존성이 존재하여 위험성이 생기게됩니다.

결국 제일 안쪽에 validation이 필수로 들어가게되겠죠. 검증을 견고하게 바꿔주세요~

JunHoPark93 replied:
실제로 작은 단위일때는 이렇게 프로젝트 코드안에서 보이지만 다른 서버들간의 통신에 있어서 validation이 누락되는 경우가 생기면 알기가 매우 힘들기때문에 최종 소비처에서는 꼭 검증이 필요합니다.

youngsu5582 replied:
다른 개발자가
우리 서버에 API 통신 했을때
IndexOutOfRangeException 같은 예외가 발생하는 걸 생각해야 하네요..!

행위 로직을 하기 전에는 검증 로직 이 필수로 들어 가야겠네요...! 감사합니다!! 😊😊

youngsu5582 reviewed (commented):
너무 부족한 시간으로 코드를 작성해서
스스로도 돌아보며 , 부족하고 아쉬운 부분이 많아서
많이 물어볼 수가 없는거 같아요... 😭😭

제이가 리뷰 해주시면 리뷰의 피드백 부분과
README TODO 리스트에 작성 한 것처럼
```
- [ ] DirectionUtil 다중 if문 개선 
- [ ] 코드 라인 10줄 이상의 코드 수정
- [ ] 인덴트 뎁스 2이상 코드 수정
- [ ] getPoint 사용 방향 수정 
- [ ] 전반적인 변수명 & 함수명 수정
- [ ] 테스트 코드 DisplayName & 테스트 코드 리팩토링
```
해당 부분 과 미구현 부분을 중점적으로 리팩토링 하겠습니다! 감사합니다 제이!! 🫡🫡

JunHoPark93 commented on src/main/java/domain/ChessBoard.java:
```diff
@@ -0,0 +1,42 @@
+        final Point endPoint = dto.getEndPoint();
+        final Piece piece = findPieceByPoint(startPoint);
+
+        if (pieces.check(piece, endPoint)) {
+            pieces.move(piece, endPoint);
+        }
+        throw new IllegalArgumentException(
+                String.format("%s 는 %s 에서 %s로 이동할 수 없습니다.", piece.getStatus(), startPoint, endPoint));
+    }
```
게임이 진행이 안되는거같네요~

youngsu5582 replied:
앗,, return 구문을 못 넣었네요..

JunHoPark93 commented on src/main/java/domain/piece/Piece.java:
```diff
@@ -0,0 +1,68 @@
+
+import domain.piece.attribute.Color;
+import domain.piece.attribute.point.Point;
+import domain.piece.kind.PieceStatus;
```
<img width="303" alt="스크린샷 2024-03-22 오후 9 50 44" src="https://github.com/woowacourse/java-chess/assets/20608121/3a88cf56-d0ea-4e3c-b676-b9df4b51d3e3">

패키지 구조가 너무 복잡합니다.
piece -> attribute -> point 까지의 depth가 필요할까요? 
추가로 piece(기물)의 attribute(속성) 에 Index, Point 등이 포함되는게 맞을까요?

그리고 kind라는 패키지 네이밍도 불필요해보여요. 간소화해주시면 좋을거같아요~

youngsu5582 replied:
도메인 을 최대한 세부 도메인으로 잘게 쪼개고 패키지로 의도 를 드러내는 것이 좋다 생각했는데
너무 구체적인 패키지 구조가
가독성 및 코드 관리에 더 어려움을 줄 수 있겠네요!

JunHoPark93 replied:
넵 처음부터 패키지를 여러 depth로 나누지말고 일단은 한군데 몰아보고 한 depth씩 나눠보면 더 나누기 편하실거에요.

JunHoPark93 commented on src/main/java/domain/piece/attribute/point/Direction.java:
```diff
@@ -0,0 +1,33 @@
+    DOWN_LEFT(index -> new Index(index.prevVertical(), index.prevHorizontal())),
+    LEFT(index -> new Index(index.vertical(), index.prevHorizontal())),
+    UP_LEFT(index -> new Index(index.nextVertical(), index.prevHorizontal()));
+    private UnaryOperator<Index> function;
```
```suggestion
    private final UnaryOperator<Index> function;
```

JunHoPark93 commented on src/main/java/domain/piece/attribute/point/Direction.java:
```diff
@@ -0,0 +1,33 @@
+        return function.apply(index);
+    }
+
+    // TODO : PAWN 에서 사용하기 위함 + 함수명이 Straight 일시 왼쪽 오른쪽 선도 생각 가능
+    public boolean isStraight() {
```
todo 주석은 개선 예정이실까요?

youngsu5582 replied:
코드 마무리 시간이 부족해서
마지막에 , 수정해야 할 부분 TODO 로 남겼는데
앞으로는 코드 외적으로 남겨야 
읽거나 보는 사람이 더 파악하기 쉬울 거 같다고 생각이 들었습니다 🥲🥲

JunHoPark93 commented on src/main/java/util/DirectionUtil.java:
```diff
@@ -0,0 +1,98 @@
+        throw new UnsupportedOperationException("생성할 수 없습니다.");
+    }
+
+    public static Direction determineDirection(final Point standard, final Point comparison) {
+        if (isUp(standard, comparison)) {
+            return Direction.UP;
+        }
+        if (isDown(standard, comparison)) {
+            return Direction.DOWN;
+        }
+        if (isLeft(standard, comparison)) {
+            return Direction.LEFT;
+        }
+        if (isRight(standard, comparison)) {
+            return Direction.RIGHT;
+        }
+        if (isUpRight(standard, comparison)) {
+            return Direction.UP_RIGHT;
+        }
+        if (isUpLeft(standard, comparison)) {
+            return Direction.UP_LEFT;
+        }
+        if (isDownRight(standard, comparison)) {
+            return Direction.DOWN_RIGHT;
```
미리 말씀주신것처럼 다수의 if문들을 개선해보시면 좋을거같아요~

youngsu5582 replied:
해당 부분은 수직 / 수평 / 대각선으로 나눈후
왼쪽 - 오른쪽 , 위 - 아래 를 통해 재분류 해서 리팩토링 했습니다!

JunHoPark93 commented on src/main/java/util/DirectionUtil.java:
```diff
@@ -0,0 +1,98 @@
+package util;
+
+import domain.piece.attribute.point.Direction;
+import domain.piece.attribute.point.Point;
+
+
+public class DirectionUtil {
+    private DirectionUtil() {
+        throw new UnsupportedOperationException("생성할 수 없습니다.");
```
DirectionUtil 은 domain패키지 하위의 Point에서만 사용되고있어요.
해당 패키지로옮겨주세요~

JunHoPark93 replied:
그리고 패키지구조를 상세하게 나눠보는건 잘못된 방향이 아니지만 역할에 맞게끔 나눠보는게 중요합니다. 참고로 종이에 그림을 그려서 역할을 나눠보면 패키지구조를 어떻게 작성할지 잘보이기도합니다.

youngsu5582 replied:
패키지를 단순
domain / util / controller / service 이렇게 나누는게 아닌
도메인 단위와 클래스의 역활로 생각하라..! 
저도 단순 기능별 분류 보다 해당 접근이 맞는거 같습니다. 감사합니다!! ☺️

JunHoPark93 commented on src/main/java/domain/piece/attribute/point/Rank.java:
```diff
@@ -0,0 +1,35 @@
+    SIX(6),
+    SEVEN(7),
+    EIGHT(8);
+    private int value;
```
```suggestion
    private final int value;
```

JunHoPark93 commented on src/main/java/domain/piece/attribute/point/Rank.java:
```diff
@@ -0,0 +1,35 @@
+            if (rank.value == value) {
+                return rank;
+            }
+        }
```
2depth를 없애볼까요?

JunHoPark93 commented on src/main/java/domain/piece/Pieces.java:
```diff
@@ -0,0 +1,76 @@
+                    .findAny();
+    }
+
+    public boolean check(final Piece piece, final Point endPoint) {
```
어떤걸 check하는지 메서드에서 명시적으로 드러내주세요

youngsu5582 replied:
해당 부분은 외부에서 check 를 한번 호출 후
move 를 호출하려 했으나
결국 move 할때도 check 가 필요한 점을 생각해서
piece 에게 canMove 책임을 주고
pieces 에서는 제거했습니다!

JunHoPark93 commented on src/main/java/domain/piece/Pieces.java:
```diff
@@ -0,0 +1,76 @@
+        if (!piece.canMove(endPoint)) {
+            return false;
+        }
+        return switch (piece.getStatus()) {
+            case BISHOP, ROOK, QUEEN -> !(hasAnyPiece(piece.getPoint(), endPoint) && !isFriend(piece, endPoint));
+            case KING, KNIGHT -> !isFriend(piece, endPoint);
+            case PAWN ->
+                    (piece.isDirectionStraight(endPoint) && !isInPiece(endPoint)) || (piece.isDirectionDiagonal(endPoint) && !isFriend(piece, endPoint));
+        };
+    }
```
switch문을 쓰지 않고 해볼수있을까요?

추가로 case문 본문의 여러 조건들이 있는데 해당 조건들을 묶어서 하나의 메서드로 만들어서 가독성을 높여볼수도있을거같아요.

youngsu5582 replied:
해당 부분은 결국
다형성 을 해치는 코드라 생각이 들어서 전부 제거했습니다!
대신 , 각 기물 들이 내부에서 해당 코드들을 호출해서
스스로가 검증 후 이동 가능한지만 반환하게 했습니다!!

JunHoPark93 replied:
잘해보셨습니다!

JunHoPark93 commented on src/main/java/domain/piece/Pieces.java:
```diff
@@ -0,0 +1,76 @@
+        return value.size();
+    }
+
+    public Map<Point, Piece> toMap() {
+        Map<Point, Piece> map = new HashMap<>();
```
자료구조가 드러나는 메서드명보다 명시적인 메서드이름으로 개선해보시면 좋을거같아요.

youngsu5582 replied:
해당 부분도 밑에 답변과 같이
도메인이 담당할 기능이 아니라고 생각해서 제거했습니다!

JunHoPark93 commented on src/main/java/domain/piece/kind/Pawn.java:
```diff
@@ -0,0 +1,65 @@
+import static domain.piece.attribute.point.Direction.*;
+
+public class Pawn extends Piece {
+    private static List<Direction> blackList = List.of(DOWN, DOWN_LEFT, DOWN_RIGHT);
```
전체적으로 final이 빠졌는데 추가해보시면 좋을거같아요.

youngsu5582 replied:
급하게 한다고 누락된 부분이 많았네요,,
Code Inspection 에서 final 키워드 Error 켜서
전부 수정했습니다!

JunHoPark93 commented on src/main/java/domain/piece/kind/Pawn.java:
```diff
@@ -0,0 +1,65 @@
+    private static final int DOUBLE_COUNT = 2;
+    private static final int SINGLE_COUNT = 1;
+
+    private boolean flag = false;
```
항상 false인데 인스턴스변수로 있어야되는걸까요? 🤔

youngsu5582 replied:
해당 부분은
2칸 이동을 하기 위해 변수는 선언했으나
시간부족으로 구현을 못했었습니다,,
추가로 변수를 통해 2칸 이동을 구현할 까 했으나

처음 만 이동 가능하다는 점 + Piece 는 자신이 이동 가능한지만 알려 준다는 점들을 통해
상수로 처음 위치를 가지고 있게 한 후 처음 위치 에서만 이동을 가능하게 했습니다!

JunHoPark93 commented on src/main/java/domain/piece/kind/Queen.java:
```diff
@@ -0,0 +1,22 @@
+    }
+
+    //TODO : 현재 방향을 계산해야만 검증 로직 가능 좀 더 생각해봐야함
+    public boolean canMove(final Point point) {
```
이 부분들도 고민해보시고 리팩토링 부탁드려요!

JunHoPark93 commented on src/main/java/domain/piece/attribute/point/File.java:
```diff
@@ -0,0 +1,35 @@
+            if (file.value == value) {
+                return file;
+            }
+        }
```
2depth도 전부 개선해보아요

JunHoPark93 reviewed (changes requested):
안녕하세요 조이썬~
힘든 미션인데 구현하느라 고생많으셨어요.
먼저 기본적인 게임동작 관련해서 확인해주세요.
코멘트 남겨드린것만 보더라도 시간이 꽤 걸릴거같은데, 일단 조이썬이 개선하고싶다고 했던 부분들먼저 봐주시고 코멘트 반영부탁드려요~ 천천히 고민해보세요.

(제 회고를 보셨다니.. 저도 제가쓴걸 4년만에 처음봤는데 진짜 꼰대같네요)

youngsu5582 committed:
refactor(All) : final 키워드 추가

[관련 피드백 링크](https://github.com/woowacourse/java-chess/pull/721#discussion_r1535542728)

youngsu5582 committed:
refactor(File,Rank) : 2 depth 제거

[관련 피드백 링크](https://github.com/woowacourse/java-chess/pull/721#discussion_r1535532063)

youngsu5582 committed:
refactor(ChessBoard) : DTO 관련 로직 도메인에서 분리

[관련 피드백 링크](https://github.com/woowacourse/java-chess/pull/721#discussion_r1535541988)

youngsu5582 committed:
feat(File,Rank) : File 과 Rank 가 메시지에 스스로 응답하게 로직 추가

File - moveLeft , moveRight , canMoveLeft , canMoveRight
Rank - moveUp , moveDown , canMoveUp , canMoveDown

youngsu5582 committed:
feat(Point) : Point 가 스스로 이동하는 기능 구현

- 이에 따른 테스트 코드 추가

youngsu5582 committed:
refactor(Point) : PointFixture 추가

테스트 코드 가독성 & 편의성 위해 Point 에 대한 Fixture 생성
- 이에 따른 테스트 코드 수정

youngsu5582 committed:
refactor(Index) : 인덱스 도메인 삭제

Index 를 사용하고 있는 코드 전부 수정
- 이에 따른 테스트 코드 수정

youngsu5582 committed:
refactor(File,Rank) : 에러 메시지에 String format 추가

youngsu5582 committed:
feat(Point,Direction) : 나이트가 이동하는 방향도 추가

UP_UP_LEFT , UP_UP_RIGHT
RIGHT_UP_RIGHT , RIGHT_DOWN_RIGHT
DOWN_DOWN_LEFT , DOWN_DOWN_RIGHT
LEFT_UP_LEFT , LEFT_DOWN_LEFT
Direction 에 방향 추가
Point 에도 관련 move , can 기능 구현 로직

youngsu5582 committed:
refactor(Knight) : 나이트 추가된 이동 방향 통해 리팩토링

youngsu5582 committed:
feat(Piece) : 일급 컬렉션에서 하는 검사 기능 객체로 이동

포인트까지 경로에 어떤 기물이든 존재하는지 검사하는 hasAnyPieceInPath
포인트에 어떤 기물이든 존재하는지 검사하는 hasAnyPiece
포인트에 아군 기물인 존재하는지 검사하는 hasFriendPiece

- 이에 따른 테스트 코드 추가

youngsu5582 committed:
feat(Piece) : 각 기물들이 스스로 기물들 사이에서 이동 가능 판단 기능 구현

Movable 인터페이스 메소드에 매개변수 추가
PiecesTest 에서 각 기물들로 테스트 분리
Pieces 에서 불필요한 기능 삭제

youngsu5582 committed:
feat(Pieces,ChessBoard) : check 후 move 가 아닌 내부적으로 check 하게 기능 구현

youngsu5582 committed:
feat(Controller) : 예외 발생시 재처리 기능 구현

youngsu5582 committed:
refactor(DirectionUtil) : if문 공통 로직 통한 간소화

[관련 피드백 링크](https://github.com/woowacourse/java-chess/pull/721#discussion_r1535528457)

youngsu5582 committed:
feat(Pawn) : 폰이 두칸 이동하는 기능 재구현

Direction 에 canMovePoint 반대인 canNotMovePoint 추가
- 이에 따른 테스트 코드 추가

youngsu5582 committed:
feat(ChessBoard) : 현재 차례와 다른 색깔의 기물 선택시 예외 기능 구현

처음 시작 색깔 포함한 생성자 추가
- 이에 따른 테스트 코드 추가

youngsu5582 committed:
fix,test(Piece,Pawn) : 존재하지 않는지 anyMatch -> allMatch 로 변경

폰이 두칸 이동 할 때 가로막혀 있을시 이동 불가능 테스트 코드 추가

youngsu5582 committed:
feat(King,Pawn) : 이동 할떄 가능한지 검증하는 로직 추가

[관련 피드백 링크](https://github.com/woowacourse/java-chess/pull/721#discussion_r1535569104)

youngsu5582 committed:
feat(Controller) : IllegalArgument , IllegalState 만 catch 하게 변경

youngsu5582 committed:
refactor(all) : 패키지 간소화

[관련 피드백 링크](https://github.com/woowacourse/java-chess/pull/721#discussion_r1535525795)

youngsu5582 committed:
refactor(Direction) : 단순 직선이 아닌 , 수직선의 의미를 드러내게 함수명 변경

[관련 피드백 링크](https://github.com/woowacourse/java-chess/pull/721#discussion_r1535526643)

youngsu5582 committed:
refactor(ChessBoard) : 불필요한 기능 제거 , 테스트 코드 삭제

findPieceByPoint 는 기본 보드판을 테스트 하기 위해서만 필요
테스트를 위한 코드라 생각하여 삭제
기본 보드판 테스트 하는 코드 역시 삭제
( 도메인 로직이 아닌 Util 에 대한 테스트 )

youngsu5582 committed:
refactor(DirectionUtil) : util 패키지가 아닌 domain 패키지로 이동

[관련 피드백 링크](https://github.com/woowacourse/java-chess/pull/721#discussion_r1535530149)

youngsu5582 committed:
refactor(All) : 상수명 재정의 & 함수 순서 가독성 있게 변경

youngsu5582 committed:
refactor(Piece) : 한번 이동하는 기능 Piece 에서 공통적으로 담당하게 변경

youngsu5582 commented:
제이!
기존의 코드와 구조가 많이 달라져서
설명을 하고 질문을 하겠습니다!
### 변경점

#### Index 제거 & Point 가 모든 방향에 대한 행위 추가
기존 Index 는  결국 Point 가 Enum Enum 으로 지정 되며 범위를 벗어나는걸 피하기 위해
만든 편법의 도메인에 불과 하다고 생각이 들었습니다...!

그래서 대신
```java
    public boolean canMoveUp(final int count) {
        return this.rank.canMoveUp(count);
    }
    public Point moveUp(final int count) {
        return new Point(this.file, this.rank.moveUp(count));
    }
```
각 방향 마다 검증 과 행위 로직을 모두 추가했습니다.

#### 이동 가능 로직 Piece 가 모두 담당

기존에는 Piece 는 자신이 해당 좌표까지 이동 가능한지 검사
-> Pieces 가 해당 기물이 다른 기물들 방해를 받지 않고 이동 가능한지 검사
이렇게 두 가지를 나눴습니다.

하지만 이런 방식 때문에
```java
case BISHOP, ROOK, QUEEN -> !(hasAnyPiece(piece.getPoint(), endPoint) && !isFriend(piece, endPoint));
case KING, KNIGHT -> !isFriend(piece, endPoint);
case PAWN ->
```
이렇게 필연적으로 다형성을 해칠수 밖에 없었습니다!

그래서 
```java
public boolean canMove(final Point movePoint, final List<Piece> piecesList) {
    final Pieces pieces = new Pieces(piecesList);
    return canMovePoint(movePoint, pieces) || canMovePointWithAttack(movePoint, pieces);
}
```
기물에 Point 와 List<Piece> 를 넘겨줘서
자체로 이동가능한지 + 기물들 사이에서 이동 가능한지에 대한 로직을 모두 수행하게 했습니다

이때 기물이 기물들 사이에서 확인해야 하는 공통된 로직 들은
Piece 라는 상위 클래스에 정의를 해놨습니다.

### 궁금한 점
#### 도메인 간 책임의 범위

체스를 리팩토링 하며 가장 주의깊게 생각했던 점은
각 도메인 들이 어디까지 검증을 하고 어디까지 수행을 해줘야 하는지에 대해서 고민을 했습니다

1. 체스 보드는 포인트가 서로 동일한지 , 해당 포인트에 기물이 있는지 , 현재 기물과 차례가 동일한지 를 담당
2. 기물 일급 컬렉션은 기물에게 이동이 가능한지 요청
3. 기물이 스스로가 모든 기물들 사이에서 이동 가능한지 요청

이렇게 협력 속에서 객체 들의 책임을 어떻게 잡아야 하는지 아직 헷갈리는거 같습니다.
특히 , 기물 일급 컬렉션에 `Point,Point` 를 넘겨주냐 , `Piece,Point` 에도 
책임과 로직 이 달라진다고 생각하기에
제이만 의 객체의 책임과 메시지를 나누는 철학이 있는지 궁금합니다!

#### 도메인의 양방향 의존?

현재 기물들의 canMove 는 List<Piece> 를 받고
해당 List 를 통해 Pieces 라는 일급 컬렉션을 만들고
상위 클래스 piece 에서 protected 접근 제한자인 공통 메소드 를 통해
갈 수 있는지 검증하고 있습니다

```java
public boolean canMove(final Point movePoint, final List<Piece> pieceList) {
    return canMovePoint(movePoint) && hasEnemyPieceOrEmpty(movePoint, new Pieces(pieceList));
}
```
```java
protected final boolean hasFriendPiece(final Point endPoint, final Pieces pieces) {
    final Optional<Piece> optionalPiece = pieces.findPieceWithPoint(endPoint);
    if (optionalPiece.isEmpty()) {
        return false;
    }
    final Piece toPiece = optionalPiece.get();
    return this.color == toPiece.color;
}
```
이렇게 Piece 내부에서 Pieces 를 생성해준후 검증 로직을 하게 구현을 했습니다.

이때 드는 생각으로
Piece 가 결국 Pieces 에 의존을 하며 서로가 양방향이 되는게 아닌가?
라는 생각이 들지만
1. Piece 가 받을때는 List<Piece> 를 통해 , 도메인이 아닌 일급 컬렉션을 받는다는 점
2. protected 를 통해 Piece 내부에서만 사용 가능한 점

해당 두 가지의 이유를 통해 Piece 가 List 를 받아서 Pieces 로 변환 후 검증 로직을 수행하게 했습니다.
위의 접근 방식이 타당한지 아니면 양방향 의존 이기에
결국 피해야 하는지에 대해서 궁금합니다! 

---

이상으로 제출을 마치겠습니다..!
생각하고 , 막히는 부분이 많아서 너무 늦게 제출한 거 같습니다,,
궁금한 부분 외에도 제이가 피드백 주고 싶거나 부족한 부분에 대해서 피드백 주시면
바로 즉각 반영하겠습니다! 감사합니다!!

( 체크 와 체크 메이트 , 프로모션 , 앙파상 같은건 구현할 수 있을지 모르겠어서 우선 제출했습니다 ,, )

JunHoPark93 commented:
> 이렇게 협력 속에서 객체 들의 책임을 어떻게 잡아야 하는지 아직 헷갈리는거 같습니다.
> 특히 , 기물 일급 컬렉션에 Point,Point 를 넘겨주냐 , Piece,Point 에도
> 책임과 로직 이 달라진다고 생각하기에
> 제이만 의 객체의 책임과 메시지를 나누는 철학이 있는지 궁금합니다!

정답은 없고 말씀주신것처럼 여러 방법이 있습니다. 철학이라기보다.. 잘게 쪼개기보다는 일단 큼직하게 만들어보고 하나씩 떼어나가는 식으로 구현하면 꼭 필요한 클래스들로만 구성되게끔 되더라구요. 변화에 대응되는 코드도 중요하지만 너무 상세하게 나눠버리게되면 관리하는 코드의 양만 늘어나게됩니다. 만들고나서는 각각이 하나의 책임을 담당하고 있는지 점검해보구요.


> Piece 가 받을때는 List 를 통해 , 도메인이 아닌 일급 컬렉션을 받는다는 점
> protected 를 통해 Piece 내부에서만 사용 가능한 점
> 해당 두 가지의 이유를 통해 Piece 가 List 를 받아서 Pieces 로 변환 후 검증 로직을 수행하게 했습니다.
> 위의 접근 방식이 타당한지 아니면 양방향 의존 이기에
> 결국 피해야 하는지에 대해서 궁금합니다!

스스로 지적해주셨군요. 넵 말씀주신것처럼 양쪽에서 사용되게 되면 실수할여지가 생기거나 코드가 복잡해집니다. 최대한 단방향으로 유지되는 구조가 훨씬 유지보수에 좋습니다. Pieces를 통째로 넘기는것이아닌 필요한 정보만 넘겨보는식으로 개선을 해볼 수 있을까요?
메서드의 인자를 정의할때도 어떤 정보들이 필요한지를 생각해보고 필요한 정보만 넘기는것이, 시그니처만 보고 어떤 일들이 일어나는지 유추가 되기 때문에 명확한 인터페이스 설계에 도움이됩니다. 한번 고민해보시고 다음 단계에서 말씀주세요!

JunHoPark93 commented:
추가적인 룰 반영보다는 현재 구조를 조금 더 고민해보시는게 더 도움될거에요!

JunHoPark93 reviewed (approved):
안녕하세요 조이썬~
리뷰 반영하느라 고생많으셨습니다.
질문주신 부분은 코멘트 남겨두었는데 다음 단계 진행하시면서 고민부탁드려요!
하시다가 잘 안되시는부분있으면 말씀주세요~
